src/ShadyCSS.js

```diff
@@ -142,33 +165,51 @@ export let ShadyCSS = {
           CS._revalidateApplyShim();
         }
         CS.applyStyles();
-        CS._documentDirty = false;
+        // if no elements have booted yet, we can just update the document and be done
+        if (!elementsHaveApplied) {
+          return;
+        }
+        // if no native css custom properties, we must recalculate the whole tree
+        if (!nativeCssVariables) {
+          this.updateStyles();
+          // make sure not to recurse back into this element subtree, unless it is the first time
+          if (hasApplied) {
+            return;
+          }
+        }
       }
     }
-    let styleInfo = StyleInfo.get(host);
-    let hasApplied = Boolean(styleInfo);
-    if (!styleInfo) {
-      styleInfo = this._prepareHost(host);
-    }
     if (overrideProps) {
       styleInfo.overrideStyleProperties =
         styleInfo.overrideStyleProperties || {};
       Object.assign(styleInfo.overrideStyleProperties, overrideProps);
     }
     if (this.nativeCss) {
+      if (styleInfo.overrideStyleProperties) {
+        this._updateNativeProperties(host, styleInfo.overrideStyleProperties);
+      }
       let template = templateMap[is];
-      if (template && template.__applyShimInvalid && template._style) {
+      // bail early if there is no shadowroot for this element
+      if (!template && is !== 'html') {
+        return;
+      }
+      if (template && template._applyShimInvalid && template._style) {
         // update template
-        ApplyShim.transformRules(template._styleAst, is);
-        template._style.textContent = StyleTransformer.elementStyles(host, styleInfo.styleRules);
+        if (!template._invalidating) {
```

Comment:
Rename `_invalidating` to `_validating` and `StyleInfo.validate` to `StyleInfo.startValidating`